pkg/cloudprovider/provider/cherryservers/provider.go

```diff
@@ -0,0 +1,429 @@
+	if s.Value == nil {
+		return nil, nil, fmt.Errorf("machine.spec.providerconfig.value is nil")
+	}
+	pconfig := providerconfig.Config{}
+	err := json.Unmarshal(s.Value.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rawConfig := RawConfig{}
+	if err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig); err != nil {
+		return nil, nil, err
+	}
+
+	c := Config{}
+	c.Token, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.Token, "CHERRYSERVERS_TOKEN")
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"token\" field, error = %v", err)
+	}
+	c.Plan, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Plan)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"plan\" field, error = %v", err)
+	}
+
+	c.ProjectID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.ProjectID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"projectID\" field, error = %v", err)
+	}
+
+	c.TeamID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.TeamID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"teamID\" field, error = %v", err)
+	}
+
+	c.Location, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Location)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"location\" field, error = %v", err)
+	}
+	c.Tags = rawConfig.Tags
+	return &c, &pconfig, err
+}
+
+func getNameForOS(os providerconfig.OperatingSystem) (string, error) {
+	switch os {
+	case providerconfig.OperatingSystemUbuntu:
+		return "Ubuntu 18.04 64bit", nil
+	case providerconfig.OperatingSystemCentOS:
+		return "CentOS 7 64bit", nil
+	}
+	return "", providerconfig.ErrOSNotSupported
+}
+
+func getClient(token string) *cherrygo.Client {
+	httpClient := &http.Client{}
+	return cherrygo.NewClientWithAuthVar(httpClient, token)
+}
+
+func getTeamByID(client *cherrygo.Client, ID string) (*cherrygo.Teams, error) {
+	teams, _, err := client.Teams.List()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, team := range teams {
+		if strconv.Itoa(team.ID) == ID {
+			return &team, nil
+		}
+	}
+	return nil, nil
+}
+
+func getPlanByName(client *cherrygo.Client, planName string, teamID string) (*cherrygo.Plans, error) {
+	tid, err := strconv.Atoi(teamID)
+	if err != nil {
+		return nil, err
+	}
+
+	plans, _, err := client.Plans.List(tid)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, plan := range plans {
+		if plan.Name == planName {
+			return &plan, nil
+		}
+	}
+	return nil, nil
+}
+
+func getServerByTag(client *cherrygo.Client, tag string, projectID string) (*cherrygo.Server, error) {
+	servers, _, err := client.Servers.List(projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, server := range servers {
+		if server.Tags[machineUIDTagKey] == tag {
+			srv, _, err := client.Server.List(strconv.Itoa(server.ID))
```

Comment:
Is it required to list the server again? Can't we just return the `server` object (i.e. `return server`)?

Reply:
For some reason I don't know cherrygo API has two different structs for representing servers: Server and Servers
Getting the list of servers with List(projectID string) ([]Servers, *Response, error) returns an array of Servers
Getting a server by ID with List(serverID string) (Server, *Response, error) returns a Server
Maybe I can unify all of that and always return a Servers object

Reply:
Oh, I see. I didn't notice that there are `Server` and `Servers`. I think it's totally okay as it is right now.